File: src/backtest.py
"""backtest.py - validate a directional setup on historical Upstox candles.

HONEST SCOPE: this backtests the UNDERLYING directional signal (the basis for an F&O
directional view), NOT option P&L. Real option P&L depends on strike, IV, theta and
path, so treat these results as evidence the *signal* has (or lacks) an edge - then
apply option structure + costs separately. PAST PERFORMANCE != FUTURE RESULTS.

Needs a valid .access_token (auth.py / get_token.py).

Examples:
    python src/backtest.py --instrument "NSE_INDEX|Nifty 50" --unit minutes --interval 5 \
        --strategy ema_crossover --fast 9 --slow 20 --stop-atr 1.5 --target-atr 3 \
        --lookback-days 30 --cost-pct 0.05

    python src/backtest.py --instrument "NSE_INDEX|Nifty Bank" --unit days --interval 1 \
        --strategy ema_crossover --lookback-days 365
"""

import os
import sys
import argparse
import datetime

import upstox_client
from upstox_client.rest import ApiException

PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
TOKEN_FILE = os.path.join(PROJECT_ROOT, ".access_token")


def load_token():
    if not os.path.exists(TOKEN_FILE):
        print(f"ERROR: {TOKEN_FILE} not found. Run auth.py or get_token.py first.")
        sys.exit(1)
    with open(TOKEN_FILE, "r", encoding="utf-8") as f:
        t = f.read().strip()
    if not t:
        print(f"ERROR: {TOKEN_FILE} is empty.")
        sys.exit(1)
    return t


def fetch_candles(client, instrument, unit, interval, lookback_days):
    """Return candles oldest-first as list of dicts: ts,o,h,l,c,v."""
    to_date = datetime.date.today().isoformat()
    from_date = (datetime.date.today() - datetime.timedelta(days=lookback_days)).isoformat()
    api = upstox_client.HistoryV3Api(client)
    resp = api.get_historical_candle_data1(instrument, unit, interval, to_date, from_date=from_date)
    raw = []
    data = getattr(resp, "data", None)
    candles = getattr(data, "candles", None) if data is not None else None
    if not candles:
        return []
    for c in candles:
        # Upstox candle: [timestamp, open, high, low, close, volume, oi]
        try:
            raw.append({
                "ts": c[0],
                "o": float(c[1]), "h": float(c[2]), "l": float(c[3]),
                "c": float(c[4]), "v": float(c[5]) if len(c) > 5 else 0.0,
            })
        except (IndexError, TypeError, ValueError):
            continue
    raw.sort(key=lambda x: x["ts"])  # oldest first
    return raw


def ema_series(values, period):
    if not values:
        return []
    k = 2.0 / (period + 1)
    out = [values[0]]
    for v in values[1:]:
        out.append(v * k + out[-1] * (1 - k))
    return out


def atr_series(candles, period=14):
    """Wilder-ish ATR; returns a list aligned to candles (first value at index 0)."""
    trs = []
    prev_close = candles[0]["c"]
    for c in candles:
        tr = max(c["h"] - c["l"], abs(c["h"] - prev_close), abs(c["l"] - prev_close))
        trs.append(tr)
        prev_close = c["c"]
    # simple rolling mean of TR as ATR proxy
    atr = []
    for i in range(len(trs)):
        window = trs[max(0, i - period + 1): i + 1]
        atr.append(sum(window) / len(window))
    return atr


def backtest_ema_crossover(candles, fast, slow, stop_atr, target_atr, cost_pct):
    closes = [c["c"] for c in candles]
    ef = ema_series(closes, fast)
    es = ema_series(closes, slow)
    atr = atr_series(candles)

    trades = []
    pos = None  # dict: side, entry, stop, target, entry_i
    for i in range(1, len(candles)):
        c = candles[i]
        crossed_up = ef[i] > es[i] and ef[i - 1] <= es[i - 1]
        crossed_dn = ef[i] < es[i] and ef[i - 1] >= es[i - 1]

        if pos is None:
            if crossed_up or crossed_dn:
                side = "LONG" if crossed_up else "SHORT"
                entry = c["c"]
                a = atr[i] or (entry * 0.001)
                if side == "LONG":
                    stop, target = entry - stop_atr * a, entry + target_atr * a
                else:
                    stop, target = entry + stop_atr * a, entry - target_atr * a
                pos = {"side": side, "entry": entry, "stop": stop, "target": target, "i": i}
            continue

        # in a position: check exits on this candle (stop before target = conservative)
        exit_price = None
        if pos["side"] == "LONG":
            if c["l"] <= pos["stop"]:
                exit_price = pos["stop"]
            elif c["h"] >= pos["target"]:
                exit_price = pos["target"]
            elif crossed_dn:
                exit_price = c["c"]
        else:
            if c["h"] >= pos["stop"]:
                exit_price = pos["stop"]
            elif c["l"] <= pos["target"]:
                exit_price = pos["target"]
            elif crossed_up:
                exit_price = c["c"]

        if exit_price is not None:
            gross = (exit_price - pos["entry"]) if pos["side"] == "LONG" else (pos["entry"] - exit_price)
            cost = (pos["entry"] + exit_price) * (cost_pct / 100.0)
            trades.append({"side": pos["side"], "entry": pos["entry"],
                           "exit": exit_price, "pnl": gross - cost})
            pos = None
    return trades


def report(trades, label):
    print("\n" + "=" * 56)
    print(f" BACKTEST RESULT - {label}")
    print("=" * 56)
    n = len(trades)
    if n == 0:
        print(" No trades generated. Try a longer lookback or different params.")
        return
    wins = [t for t in trades if t["pnl"] > 0]
    losses = [t for t in trades if t["pnl"] <= 0]
    total = sum(t["pnl"] for t in trades)
    gross_win = sum(t["pnl"] for t in wins)
    gross_loss = -sum(t["pnl"] for t in losses)
    win_rate = 100.0 * len(wins) / n
    avg_win = (gross_win / len(wins)) if wins else 0.0
    avg_loss = (gross_loss / len(losses)) if losses else 0.0
    expectancy = total / n
    profit_factor = (gross_win / gross_loss) if gross_loss > 0 else float("inf")

    # max drawdown on the equity curve (in points)
    equity, peak, max_dd = 0.0, 0.0, 0.0
    for t in trades:
        equity += t["pnl"]
        peak = max(peak, equity)
        max_dd = max(max_dd, peak - equity)

    print(f" Trades:        {n}   (wins {len(wins)}, losses {len(losses)})")
    print(f" Win rate:      {win_rate:.1f}%")
    print(f" Avg win:       {avg_win:.2f} pts   Avg loss: {avg_loss:.2f} pts")
    print(f" Expectancy:    {expectancy:.2f} pts/trade   (>0 = positive edge, net of costs)")
    print(f" Profit factor: {profit_factor:.2f}   (>1 = profitable)")
    print(f" Total:         {total:.2f} pts   Max drawdown: {max_dd:.2f} pts")
    print("=" * 56)
    print(" NOTE: underlying-signal backtest only - not option P&L. Past != future.")


def main():
    p = argparse.ArgumentParser(description="Backtest a directional setup on Upstox candles.")
    p.add_argument("--instrument", required=True, help='e.g. "NSE_INDEX|Nifty 50"')
    p.add_argument("--unit", default="minutes", help="minutes/hours/days/weeks/months")
    p.add_argument("--interval", default="5")
    p.add_argument("--strategy", default="ema_crossover", choices=["ema_crossover"])
    p.add_argument("--fast", type=int, default=9)
    p.add_argument("--slow", type=int, default=20)
    p.add_argument("--stop-atr", type=float, default=1.5)
    p.add_argument("--target-atr", type=float, default=3.0)
    p.add_argument("--lookback-days", type=int, default=30)
    p.add_argument("--cost-pct", type=float, default=0.05, help="round-trip cost as %% of price")
    args = p.parse_args()

    client = upstox_client.ApiClient(_config(load_token()))
    try:
        candles = fetch_candles(client, args.instrument, args.unit, args.interval, args.lookback_days)
    except ApiException as e:
        print(f"ERROR fetching candles: HTTP {e.status}: {e.body}")
        sys.exit(1)

    if len(candles) < max(args.slow + 5, 20):
        print(f"Only {len(candles)} candles fetched - not enough to backtest. Increase --lookback-days.")
        sys.exit(1)

    trades = backtest_ema_crossover(candles, args.fast, args.slow, args.stop_atr,
                                    args.target_atr, args.cost_pct)
    label = f"{args.instrument} {args.interval}{args.unit} EMA{args.fast}/{args.slow} " \
            f"SL{args.stop_atr}xATR TP{args.target_atr}xATR ({len(candles)} candles)"
    report(trades, label)


def _config(token):
    cfg = upstox_client.Configuration()
    cfg.access_token = token
    return cfg


if __name__ == "__main__":
    main()

File: src/main/java/com/zaheenmunshi/upstox/Backtest.java
package com.zaheenmunshi.upstox;

import com.upstox.ApiClient;
import com.upstox.ApiException;
import com.upstox.api.GetHistoricalCandleResponse;
import io.swagger.client.api.HistoryV3Api;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

/**
 * Port of {@code backtest.py}: validate a directional setup on historical Upstox candles.
 *
 * HONEST SCOPE: this backtests the UNDERLYING directional signal, NOT option P&L.
 * Treat results as evidence the *signal* has (or lacks) an edge; apply option
 * structure + costs separately. PAST PERFORMANCE != FUTURE RESULTS.
 */
public final class Backtest {

    private record Candle(String ts, double o, double h, double l, double c, double v) {}
    private record Trade(String side, double entry, double exit, double pnl) {}

    private Backtest() {}

    public static void main(String[] args) {
        // Defaults mirror the Python argparse.
        String instrument = null, unit = "minutes", strategy = "ema_crossover";
        int interval = 5, fast = 9, slow = 20, lookbackDays = 30;
        double stopAtr = 1.5, targetAtr = 3.0, costPct = 0.05;

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--instrument"    -> instrument = args[++i];
                case "--unit"          -> unit = args[++i];
                case "--interval"      -> interval = Integer.parseInt(args[++i]);
                case "--strategy"      -> strategy = args[++i];
                case "--fast"          -> fast = Integer.parseInt(args[++i]);
                case "--slow"          -> slow = Integer.parseInt(args[++i]);
                case "--stop-atr"      -> stopAtr = Double.parseDouble(args[++i]);
                case "--target-atr"    -> targetAtr = Double.parseDouble(args[++i]);
                case "--lookback-days" -> lookbackDays = Integer.parseInt(args[++i]);
                case "--cost-pct"      -> costPct = Double.parseDouble(args[++i]);
                default -> { System.out.println("Unknown argument: " + args[i]); System.exit(2); }
            }
        }
        if (instrument == null) {
            System.out.println("ERROR: --instrument is required, e.g. --instrument \"NSE_INDEX|Nifty 50\"");
            System.exit(2);
        }
        if (!strategy.equals("ema_crossover")) {
            System.out.println("ERROR: unsupported --strategy " + strategy + " (only ema_crossover).");
            System.exit(2);
        }

        Config cfg = Config.load();
        ApiClient client = UpstoxClients.authenticated(cfg.loadToken());

        List<Candle> candles;
        try {
            candles = fetchCandles(client, instrument, unit, interval, lookbackDays);
        } catch (ApiException e) {
            System.out.println("ERROR fetching candles: HTTP " + e.getCode() + ": " + e.getResponseBody());
            System.exit(1);
            return;
        }

        if (candles.size() < Math.max(slow + 5, 20)) {
            System.out.println("Only " + candles.size() + " candles fetched - not enough to backtest. Increase --lookback-days.");
            System.exit(1);
        }

        List<Trade> trades = backtestEmaCrossover(candles, fast, slow, stopAtr, targetAtr, costPct);
        String label = String.format("%s %d%s EMA%d/%d SL%sxATR TP%sxATR (%d candles)",
                instrument, interval, unit, fast, slow, stopAtr, targetAtr, candles.size());
        report(trades, label);
    }

    /** Return candles oldest-first. Upstox candle = [ts, open, high, low, close, volume, oi]. */
    private static List<Candle> fetchCandles(ApiClient client, String instrument, String unit,
                                             int interval, int lookbackDays) throws ApiException {
        String toDate = LocalDate.now().toString();
        String fromDate = LocalDate.now().minusDays(lookbackDays).toString();
        GetHistoricalCandleResponse resp =
                new HistoryV3Api(client).getHistoricalCandleData1(instrument, unit, interval, toDate, fromDate);

        List<Candle> out = new ArrayList<>();
        if (resp.getData() == null || resp.getData().getCandles() == null) return out;
        for (List<Object> c : resp.getData().getCandles()) {
            try {
                out.add(new Candle(
                        String.valueOf(c.get(0)),
                        num(c.get(1)), num(c.get(2)), num(c.get(3)), num(c.get(4)),
                        c.size() > 5 ? num(c.get(5)) : 0.0));
            } catch (RuntimeException ignored) {
                // skip malformed candle (matches Python's try/except continue)
            }
        }
        out.sort((a, b) -> a.ts.compareTo(b.ts)); // oldest first
        return out;
    }

    private static double num(Object o) { return Double.parseDouble(String.valueOf(o)); }

    private static double[] emaSeries(double[] values, int period) {
        if (values.length == 0) return new double[0];
        double k = 2.0 / (period + 1);
        double[] out = new double[values.length];
        out[0] = values[0];
        for (int i = 1; i < values.length; i++) out[i] = values[i] * k + out[i - 1] * (1 - k);
        return out;
    }

    /** Wilder-ish ATR (simple rolling mean of TR), aligned to candles. */
    private static double[] atrSeries(List<Candle> candles, int period) {
        int n = candles.size();
        double[] trs = new double[n];
        double prevClose = candles.get(0).c;
        for (int i = 0; i < n; i++) {
            Candle c = candles.get(i);
            trs[i] = Math.max(c.h - c.l, Math.max(Math.abs(c.h - prevClose), Math.abs(c.l - prevClose)));
            prevClose = c.c;
        }
        double[] atr = new double[n];
        for (int i = 0; i < n; i++) {
            int start = Math.max(0, i - period + 1);
            double sum = 0; int cnt = 0;
            for (int j = start; j <= i; j++) { sum += trs[j]; cnt++; }
            atr[i] = sum / cnt;
        }
        return atr;
    }

    private static List<Trade> backtestEmaCrossover(List<Candle> candles, int fast, int slow,
                                                    double stopAtr, double targetAtr, double costPct) {
        int n = candles.size();
        double[] closes = new double[n];
        for (int i = 0; i < n; i++) closes[i] = candles.get(i).c;
        double[] ef = emaSeries(closes, fast);
        double[] es = emaSeries(closes, slow);
        double[] atr = atrSeries(candles, 14);

        List<Trade> trades = new ArrayList<>();
        String side = null; double entry = 0, stop = 0, target = 0;

        for (int i = 1; i < n; i++) {
            Candle c = candles.get(i);
            boolean crossedUp = ef[i] > es[i] && ef[i - 1] <= es[i - 1];
            boolean crossedDn = ef[i] < es[i] && ef[i - 1] >= es[i - 1];

            if (side == null) {
                if (crossedUp || crossedDn) {
                    side = crossedUp ? "LONG" : "SHORT";
                    entry = c.c;
                    double a = atr[i] != 0 ? atr[i] : entry * 0.001;
                    if (side.equals("LONG")) { stop = entry - stopAtr * a; target = entry + targetAtr * a; }
                    else { stop = entry + stopAtr * a; target = entry - targetAtr * a; }
                }
                continue;
            }

            Double exitPrice = null;
            if (side.equals("LONG")) {
                if (c.l <= stop) exitPrice = stop;
                else if (c.h >= target) exitPrice = target;
                else if (crossedDn) exitPrice = c.c;
            } else {
                if (c.h >= stop) exitPrice = stop;
                else if (c.l <= target) exitPrice = target;
                else if (crossedUp) exitPrice = c.c;
            }

            if (exitPrice != null) {
                double gross = side.equals("LONG") ? (exitPrice - entry) : (entry - exitPrice);
                double cost = (entry + exitPrice) * (costPct / 100.0);
                trades.add(new Trade(side, entry, exitPrice, gross - cost));
                side = null;
            }
        }
        return trades;
    }

    private static void report(List<Trade> trades, String label) {
        System.out.println("\n" + "=".repeat(56));
        System.out.println(" BACKTEST RESULT - " + label);
        System.out.println("=".repeat(56));
        int n = trades.size();
        if (n == 0) {
            System.out.println(" No trades generated. Try a longer lookback or different params.");
            return;
        }
        List<Trade> wins = new ArrayList<>(), losses = new ArrayList<>();
        double total = 0, grossWin = 0, grossLoss = 0;
        for (Trade t : trades) {
            total += t.pnl;
            if (t.pnl > 0) { wins.add(t); grossWin += t.pnl; }
            else { losses.add(t); grossLoss -= t.pnl; }
        }
        double winRate = 100.0 * wins.size() / n;
        double avgWin = wins.isEmpty() ? 0 : grossWin / wins.size();
        double avgLoss = losses.isEmpty() ? 0 : grossLoss / losses.size();
        double expectancy = total / n;
        String profitFactor = grossLoss > 0 ? String.format("%.2f", grossWin / grossLoss) : "inf";

        double equity = 0, peak = 0, maxDd = 0;
        for (Trade t : trades) {
            equity += t.pnl;
            peak = Math.max(peak, equity);
            maxDd = Math.max(maxDd, peak - equity);
        }

        System.out.printf(" Trades:        %d   (wins %d, losses %d)%n", n, wins.size(), losses.size());
        System.out.printf(" Win rate:      %.1f%%%n", winRate);
        System.out.printf(" Avg win:       %.2f pts   Avg loss: %.2f pts%n", avgWin, avgLoss);
        System.out.printf(" Expectancy:    %.2f pts/trade   (>0 = positive edge, net of costs)%n", expectancy);
        System.out.printf(" Profit factor: %s   (>1 = profitable)%n", profitFactor);
        System.out.printf(" Total:         %.2f pts   Max drawdown: %.2f pts%n", total, maxDd);
        System.out.println("=".repeat(56));
        System.out.println(" NOTE: underlying-signal backtest only - not option P&L. Past != future.");
    }
}
